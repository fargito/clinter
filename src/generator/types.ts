import { Linter } from "eslint";
import { AnswerObject } from "types";

export type ESLintGenerator = (userAnswers: AnswerObject) => (config: Linter.Config) => Linter.Config;

export type ESLintDependencyGenerator = (userAnswers: AnswerObject) => (deps: string[]) => string[];
