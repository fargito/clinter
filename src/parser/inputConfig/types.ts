import { ModeInfoObject, ProjectInfoObject } from "types";

export interface InputConfig {
  generatorConfig: ProjectInfoObject;
  modeConfig: ModeInfoObject;
}
