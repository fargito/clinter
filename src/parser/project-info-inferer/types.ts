import { ProjectInfoObject } from "types";

export enum ProjectInfoRetrievalMode {
  Automatic = "automatic",
  Manual = "manual",
}

export interface ProjectInfoRetrievalModeObject {
  mode: ProjectInfoRetrievalMode;
}

export type ProjectInfoInferer = (projectInfo: ProjectInfoObject) => ProjectInfoObject;
