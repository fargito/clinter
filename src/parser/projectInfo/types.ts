import { ProjectInfoObject } from "types";

export enum ProjectInfoRetrievalMode {
  Automatic = "automatic",
  Manual = "manual",
}

export interface ProjectInfoRetrievalModeObject {
  mode: ProjectInfoRetrievalMode;
}

export type ProjectInfoParser = (projectInfo: ProjectInfoObject) => ProjectInfoObject;
