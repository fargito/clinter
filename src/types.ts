export enum FormatterAnswer {
  None = "None",
  Prettier = "Prettier",
}

export interface FormatterAnswerObject {
  formatter: FormatterAnswer;
}

export enum FrontFrameworkAnswer {
  None = "None",
  Vue = "Vue",
  React = "React",
}

export interface FrontFrameworkAnswerObject {
  frontFramework: FrontFrameworkAnswer;
}

export enum TypescriptAnswer {
  None = "None",
  WithTypeChecking = "With type checking",
  NoTypeChecking = "Without type checking",
}

export interface TypescriptAnswerObject {
  typescript: TypescriptAnswer;
}

export enum EnvAnswer {
  Browser = "Browser",
  Node = "Node",
}

export interface EnvAnswerObject {
  env: EnvAnswer[];
}

export enum TestFrameworkAnswer {
  None = "None",
  Jest = "Jest",
}

export interface TestFrameworkAnswerObject {
  test: TestFrameworkAnswer;
}

export type AnswerObject = FormatterAnswerObject &
  FrontFrameworkAnswerObject &
  TypescriptAnswerObject &
  EnvAnswerObject &
  TestFrameworkAnswerObject;
