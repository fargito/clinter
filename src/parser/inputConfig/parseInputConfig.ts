import signale from "signale";

import {
  promptGeneratorUserQuestions,
  promptModeUserQuestions,
  promptProjectInfoRetrievalModeQuestions,
} from "parser/user-questions";
import { InputConfig } from "parser/inputConfig/types";
import { ProjectInfoRetrievalMode } from "parser/projectInfo/types";

export async function parseInputConfig(): Promise<InputConfig> {
  const projectInfoRetrievalMode = await promptProjectInfoRetrievalModeQuestions();

  const modeConfig = await promptModeUserQuestions();

  const generatorConfig = await promptGeneratorUserQuestions();
  signale.info("Retrieving project settings ...");

  return {
    modeConfig,
    generatorConfig,
  };
}
