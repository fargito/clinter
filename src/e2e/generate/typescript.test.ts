import { defaultInputConfig, makeTestService } from "e2e/helpers";
import { InputConfig } from "parser/inputConfig/types";
import { TypescriptAnswer } from "types";

describe("TypeScript configuration Generator Mode", () => {
  const testService = makeTestService();

  it("should return a typescript eslint configuration with no type checking", async () => {
    const inputConfig: InputConfig = {
      ...defaultInputConfig,
      generatorConfig: {
        ...defaultInputConfig.generatorConfig,
        typescript: TypescriptAnswer.NoTypeChecking,
      },
    };

    testService.loadInputConfig(inputConfig);
    await testService.runClinter();
    const outputConfig = testService.getParsedOutputConfig();

    expect(outputConfig).toMatchSnapshot("eslint-ts-no-type-checking-config");
  });

  it("should return a typescript eslint configuration with type checking", async () => {
    const inputConfig: InputConfig = {
      ...defaultInputConfig,
      generatorConfig: {
        ...defaultInputConfig.generatorConfig,
        typescript: TypescriptAnswer.WithTypeChecking,
      },
    };

    testService.loadInputConfig(inputConfig);
    await testService.runClinter();
    const outputConfig = testService.getParsedOutputConfig();

    expect(outputConfig).toMatchSnapshot("eslint-ts-type-checking-config");
  });
});
