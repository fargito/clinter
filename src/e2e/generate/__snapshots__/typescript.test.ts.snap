// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeScript configuration Generator Mode should return a typescript eslint configuration with no type checking: eslint-ts-no-type-checking-config 1`] = `
Object {
  "env": Object {},
  "extends": Array [
    "eslint:recommended",
    "plugin:@typescript-eslint/recommended",
  ],
  "parser": "@typescript-eslint/parser",
  "parserOptions": Object {
    "project": "tsconfig.json",
  },
  "plugins": Array [
    "import",
  ],
  "root": true,
  "rules": Object {
    "@typescript-eslint/prefer-optional-chain": "error",
    "complexity": Array [
      "error",
      8,
    ],
    "curly": Array [
      "error",
      "all",
    ],
    "eqeqeq": Array [
      "error",
      "smart",
    ],
    "import/no-extraneous-dependencies": Array [
      "error",
      Object {
        "devDependencies": true,
        "optionalDependencies": false,
        "peerDependencies": false,
      },
    ],
    "import/order": Array [
      "error",
      Object {
        "groups": Array [
          Array [
            "external",
            "builtin",
          ],
          "internal",
          Array [
            "parent",
            "sibling",
            "index",
          ],
        ],
      },
    ],
    "no-shadow": Array [
      "error",
      Object {
        "hoist": "all",
      },
    ],
    "prefer-const": "error",
    "sort-imports": Array [
      "error",
      Object {
        "ignoreCase": true,
        "ignoreDeclarationSort": true,
        "ignoreMemberSort": false,
        "memberSyntaxSortOrder": Array [
          "none",
          "all",
          "multiple",
          "single",
        ],
      },
    ],
  },
}
`;

exports[`TypeScript configuration Generator Mode should return a typescript eslint configuration with type checking: eslint-ts-type-checking-config 1`] = `
Object {
  "env": Object {},
  "extends": Array [
    "eslint:recommended",
    "plugin:@typescript-eslint/recommended",
    "plugin:@typescript-eslint/recommended-requiring-type-checking",
  ],
  "parser": "@typescript-eslint/parser",
  "parserOptions": Object {
    "project": "tsconfig.json",
  },
  "plugins": Array [
    "import",
  ],
  "root": true,
  "rules": Object {
    "@typescript-eslint/prefer-nullish-coalescing": "error",
    "@typescript-eslint/prefer-optional-chain": "error",
    "@typescript-eslint/strict-boolean-expressions": "error",
    "complexity": Array [
      "error",
      8,
    ],
    "curly": Array [
      "error",
      "all",
    ],
    "eqeqeq": Array [
      "error",
      "smart",
    ],
    "import/no-extraneous-dependencies": Array [
      "error",
      Object {
        "devDependencies": true,
        "optionalDependencies": false,
        "peerDependencies": false,
      },
    ],
    "import/order": Array [
      "error",
      Object {
        "groups": Array [
          Array [
            "external",
            "builtin",
          ],
          "internal",
          Array [
            "parent",
            "sibling",
            "index",
          ],
        ],
      },
    ],
    "no-shadow": Array [
      "error",
      Object {
        "hoist": "all",
      },
    ],
    "prefer-const": "error",
    "sort-imports": Array [
      "error",
      Object {
        "ignoreCase": true,
        "ignoreDeclarationSort": true,
        "ignoreMemberSort": false,
        "memberSyntaxSortOrder": Array [
          "none",
          "all",
          "multiple",
          "single",
        ],
      },
    ],
  },
}
`;
